File: cpp/src/day1.cc
#include "aoc.hh"
#include <cctype>
#include <cstring>
#include <iostream>
#include <sstream>
#include <string>

int day1p1(const std::string& input) {
    std::istringstream iss(input);
    std::string line;
    int res = 0;
    while (std::getline(iss, line, '\n')) {
        int num1 = -1;
        int num2 = -1;
        for (auto& ch : line) {
            if (isdigit(ch)) {
                if (num1 == -1) {
                    num1 = ch - 48;
                }
                num2 = ch - 48;
            }
        }
        int num = (num1 * 10) + num2;
        res += num;
    }
    return res;
}

struct lookup {
    const char* str;
    size_t str_len;
    int num;
};

lookup lookups[] = {
    {"one", 3, 1},   {"two", 3, 2},   {"three", 5, 3},
    {"four", 4, 4},  {"five", 4, 5},  {"six", 3, 6},
    {"seven", 5, 7}, {"eight", 5, 8}, {"nine", 4, 9},
};

size_t lookups_len = sizeof lookups / sizeof lookups[0];

int lookup_str(const char* str) {
    size_t i;
    for (i = 0; i < lookups_len; ++i) {
        if (std::strncmp(lookups[i].str, str, lookups[i].str_len) == 0) {
            return lookups[i].num;
        }
    }
    return -1;
}

int day1p2(const std::string& input) {
    std::istringstream iss(input);
    std::string line;
    int res = 0;
    while (std::getline(iss, line, '\n')) {
        int num1 = -1;
        int num2 = -1;
        size_t i, len = line.length();
        for (i = 0; i < len; ++i) {
            std::string slice = line.substr(i, len - i);
            int lookup = lookup_str(slice.c_str());
            if (lookup == -1) {
                char ch = line[i];
                if (isdigit(ch)) {
                    if (num1 == -1) {
                        num1 = ch - 48;
                    }
                    num2 = ch - 48;
                }
            } else {
                if (num1 == -1) {
                    num1 = lookup;
                }
                num2 = lookup;
            }
        }
        int num = (num1 * 10) + num2;
        res += num;
    }
    return res;
}

File: js/src/day1.js
import { read } from "./reader.js";

const numCodes = {
    zero: 48,
    nine: 57,
};

/**
 * @param {string} input
 * @returns {number} result
 */
export function p1(input) {
    let lines = input.split('\n').filter((line) => line != '');
    let res = 0;
    /** @type {number} */
    let i;
    for (i = 0; i < lines.length; ++i) {
        let line = lines[i];
        /** @type {number} */
        let j;
        let state = "";
        for (j = 0; j < line.length; ++j) {
            let ch = line[j];
            if (numCodes.zero <= ch.charCodeAt(0) && ch.charCodeAt(0) <= numCodes.nine) {
                state += ch;
                break;
            }
        }
        for (j = line.length - 1; j >= 0; j--) {
            let ch = line[j];
            if (numCodes.zero <= ch.charCodeAt(0) && ch.charCodeAt(0) <= numCodes.nine) {
                state += ch;
                break;
            }
        }
        let num = parseInt(state);
        res += num;
    }
    return res;
}

const lookups = [
    { str: "one", num: 1 },
    { str: "two", num: 2 },
    { str: "three", num: 3 },
    { str: "four", num: 4 },
    { str: "five", num: 5 },
    { str: "six", num: 6 },
    { str: "seven", num: 7 },
    { str: "eight", num: 8 },
    { str: "nine", num: 9 },
];

/**
 * @param {string} str
 * @returns {number} lookup num
 */
function lookup(str) {
    /** @type number */
    let i;
    for (i = 0; i < lookups.length; ++i) {
        let lookup = lookups[i];
        if (str.startsWith(lookup.str)) {
            return lookup.num;
        }
    }
    return -1;
}

/**
 * @param {string} input
 * @returns {number} result
 */
export function p2(input) {
    let lines = input.split('\n').filter((line) => line != '');
    let res = 0;
    /** @type {number} */
    let i;
    for (i = 0; i < lines.length; ++i) {
        let line = lines[i];
        /** @type {number} */
        let j;

        let num1 = -1;
        let num2 = -1;
        for (j = 0; j < line.length; ++j) {
            let sub = line.substring(j);
            let numFromSub = lookup(sub);
            if (numFromSub === -1) {
                let ch = line[j];
                if (numCodes.zero <= ch.charCodeAt(0) && ch.charCodeAt(0) <= numCodes.nine) {
                    if (num1 == -1) {
                        num1 = ch.charCodeAt(0) - 48;
                    }
                    num2 = ch.charCodeAt(0) - 48;
                }
            } else {
                if (num1 == -1) {
                    num1 = numFromSub;
                }
                num2 = numFromSub;
            }
        }
        let num = (num1 * 10) + num2;
        res += num;
    }
    return res;
}

async function main() {
    let input = await read(null);
    let res = p2(input);
    console.log({ res });
}

main();

